### ISSUE TITLE
what optimized hyperparameters for Ising 32*32

### ISSUE THREAD

**Issue by shijiale0609** [lherron2/thermomaps-ising#2]
It is a great work. I want to first reproduce it and adopt it to my own studying system.

from the provided [jupyter notebook](https://github.com/lherron2/thermomaps-ising/blob/main/thermomaps_ising.ipynb)
``````
model = Unet2D(dim=16, dim_mults=(1, 2, 4), resnet_block_groups=8, channels=2)
backbone = ConvBackbone(model=model,
                        data_shape=train_loader.data_dim,
                        target_shape=8,
                        num_dims=4,
                        lr=1e-3,
                        eval_mode="train")

diffusion = VPDiffusion(num_diffusion_timesteps=100)


trainer = DiffusionTrainer(diffusion,
                           backbone,
                           train_loader,
                           prior,
                           model_dir="/content/models", # save models every epoch
                           pred_type="noise", # set to "noise" or "x0"
                           test_loader=test_loader # optional
                           )
     
trainer.train(20, loss_type="smooth_l1", batch_size=8)
``````

The given notebook is about 8*8 matrix. When I run the notebook, I can approximately reproduce the results (slightly different from the results in the notebook)
.
The PNAS paper is about 32*32 matrix. I tried to reproduce the PNAS paper results by adapting the notebook. However, the generated results is quite away from the simulated results and the results in the PNAS paper. 

``````
model = Unet2D(dim=32, dim_mults=(1, 2, 4), resnet_block_groups=8, channels=2)
backbone = ConvBackbone(model=model,
                        data_shape=train_loader.data_dim,
                        target_shape=32,
                        num_dims=4,
                        lr=1e-3,
                        eval_mode="train")

diffusion = VPDiffusion(num_diffusion_timesteps=100)


trainer = DiffusionTrainer(diffusion,
                           backbone,
                           train_loader,
                           prior,
                           model_dir="/content/models", # save models every epoch
                           pred_type="noise", # set to "noise" or "x0"
                           test_loader=test_loader # optional
                           )
     
trainer.train(500, loss_type="smooth_l1", batch_size=64)
``````

However, the results are quite away from the results shown in the PNAS paper. 

<img width="1800" height="600" alt="Image" src="https://github.com/user-attachments/assets/b59661e0-ad13-4048-94a1-0d4886baa068" />

I wonder what are the hyperparameters that you used? Could you provide them to help me reproduce the results?

**Comment by lherron2** [comment]
Hi Jiale,

Thanks for taking an interest in our work. I will work on the issue in the coming days. For the interactive example I wanted something that would run in a Colab notebook in ~10ish minutes, hence the focus on 8x8 model. 

In the meanwhile, the main trick to scaling up to 32x32 is training the network on more samples at each training temperature. We think that the TM model learns the temperature dependence from fluctuations in the training samples (i.e., fluctuations in the Ising spins at a given temperature), so as the size of the Ising model increases so too does the amount of data needed to train the model. How many samples are you training on?

In any case, I will check the settings we used for the neural net in the PNAS and tinker with the demo to get it working in the 32x32 case.

**Comment by shijiale0609** [comment]
Hi Lukas, Thanks for your response.

I directly adapt the given code to generate 5000 samples for 32x32 case. 
``````
  IM = IsingModel(sampler=SwendsenWangSampler, size = 32, warmup = 1000, temp = np.round(temperature,1), Jx = 1, Jy = 1)
  IM.simulate(steps = 5000, observables = [Energy(Jx = 1, Jy = 1), Magnetization()], sampling_frequency = 1)
``````

**Comment by lherron2** [comment]
Hi Jiale,

Here are the parameters that you can use to approximately reproduce the 16x16 Ising result. In short, use the SW sampler to generate ~100,000 lattice configurations:
```
IM = IsingModel(sampler=SwendsenWangSampler, size=16, warmup=100, temp=np.round(temperature, 1), Jx=1, Jy=1)
IM.simulate(steps=500_000, observables=[Energy(Jx=1, Jy=1), Magnetization()], sampling_frequency=5)
```
Then train a model that is larger than the one provided with the original notebook with a small learning rate and a large batch size for as long as you can afford.
```
model = Unet2D(dim=64, dim_mults=(1, 2, 4, 8), resnet_block_groups=2, channels=2)
backbone = ConvBackbone(..., lr=1e-5, ...)
trainer = DiffusionTrainer(...)
trainer.train(300, loss_type="smooth_l1", batch_size=64)
```

I tried using these parameters to train a 32x32 Ising model but was not able to get the model's training to converge. Training oscillated around/close to the right generalization. See the inline screenshots taken one epoch apart.

<img width="863" height="419" alt="Image" src="https://github.com/user-attachments/assets/0ed1dee3-dcde-444f-939a-ce75c5c2f508" />

<img width="870" height="425" alt="Image" src="https://github.com/user-attachments/assets/03333142-57bf-4386-bb29-07a9a0a65b79" />

Then I stepped down to 16x16 keeping the rest of the parameters the same. The results looked much better, although I didn't wait for training to fully converge. Each of these were each taken several epochs (~1hour wall clock time) apart.

<img width="861" height="414" alt="Image" src="https://github.com/user-attachments/assets/85b5d4e9-d829-417a-9717-562e44beca88" />

<img width="864" height="416" alt="Image" src="https://github.com/user-attachments/assets/412b9af7-8676-47e6-a0df-10cc9f2b2175" />

<img width="867" height="416" alt="Image" src="https://github.com/user-attachments/assets/c84015e2-1463-412d-937e-f3c40924f31e" />

If you plan to work with these models yourself I will give you some tips on how to train them, because it is difficult to get them to converge for these sort of generalization tests. 

First, the quality of the training data is highly important for getting the model to converge, and it becomes increasingly important as the size of the system increases.  We have reasoned that the model's generalization capacity improves the more it is trained on samples that energetically 'look' like ones from the transition region, and these become increasingly rare as (i) one shifts the simulation temperature away from the critical point and (ii) increases the system size $N$. At a given temperature the model samples an energy distribution with $\langle U \rangle = N\cdot k_BT$ and $\mathrm{Var}(U)\propto N$. The relative fluctuations in the energy then scale as $\sqrt{\mathrm{Var(U)}} / {\langle U \rangle} \propto N^{-1/2}$. As the system size increases the relative magnitude of the fluctuations decreases as $N^{-1/2}$. Across all of the works that have used thermodynamic maps [[a](https://www.pnas.org/doi/abs/10.1073/pnas.2321971121),[b](https://arxiv.org/abs/2503.15080),[c](https://journals.aps.org/prl/abstract/10.1103/1wdj-ym3s)] we have found that sufficient sampling at the training temperatures is required, and one should take the distance from the critical point and the system size into account. For the 32x32 Ising model I recall generating >100,000 samples on the HPC at each of the training temperatures.

The second bit of advice concerns training the model. Since TM learns to generalize from relative fluctuations in the energy distribution, we have to ensure that those signal consistently appear in the batches that the model is trained on. Hence, one should use a large batch size. But as the batch size increases, the training becomes less stochastic and more like convex gradient descent, so one must also decrease the learning rate. The combination of a large batch size and a small learning rate means that one must train for a very long time. In the PNAS I used a batch size of 512 for the 32x32 model and annealed the learning rate to 1e-7. I do not have saved the number of epochs that I trained for, but I recall the wall time being very long and essentially maxing out the memory on an 80GB A100 between the model parameters and the batch size. If you want to simulate a larger batch size on more memory-limited hardware you can use `trainer.train(300, loss_type="smooth_l1", batch_size=128,  grad_accumulation_steps=10)` to emulate a larger batch size.

**Comment by shijiale0609** [comment]
Thanks Lukas for your detailed explanation. I was trying to do and train the model as you suggested.

I still have several other questions and hope that you can give me some suggestions. 
1. How do you set and optimize the hyperparameters of the Unet 2D. 
  For size 32 * 32, you set model = Unet2D(dim=64, dim_mults=(1, 2, 4, 8), resnet_block_groups=2, channels=2). the example size 8 * 8 in the notebook, the model set up is different. 

2.  Why 1.5 and 2.5, these two specific temperatures are chosen? Do you try 1.3, 2.7 instead of 1.5, 2.5? 
3. Have you tried to increase the training data, from  two temperature 1.5, 2.5 to four temperature, 1.3, 1.5, 2.5, 2.7, for even more?
